### DISCUSSION TITLE
Error: connect ECONNREFUSED

### DISCUSSION THREAD

**Discussion by ghost** [sidorares/node-mysql2#1566] [Q&A] [answered]
Hey, I just created a mysql server in unbutu and an user for externals server connect to the database : 

Database :
![image](https://user-images.githubusercontent.com/86203108/165366004-25f5a840-81cd-4771-a421-d1aa0d35a0ac.png)

User :
![image](https://user-images.githubusercontent.com/86203108/165365955-70a494a2-7418-4de8-b0a5-562f94118522.png)
(I don't know why, but the authentication_string is very different than the password that I put for the user and that I'm using to connect with mysql2 on nodejs).

Nodejs code : 

```js
const db = createConnection({
	user: "gameServer",
	database: "wildDB",
	host: "serverIP",
	port: 3306,
	password: "myPassword",
});
```

(I checked, and the server is running on port 3306)

And here is the error that I get : 
![image](https://user-images.githubusercontent.com/86203108/165366681-bea67f69-02aa-4426-8368-c6b1e3add8e4.png)

Any idea to fix it ? (It's maybe a stupid error, it's the first time I use mysql)

**Comment by sidorares** [comment]
can you double check the server is listening on TCP interface? mysql cli might be using a unix socket to connect by default

**Reply by ghost** [reply]
Do you mean that ?

![image](https://user-images.githubusercontent.com/86203108/165435251-e0f55dc9-7fc5-4d5b-bf0d-40cf175bdae5.png)

**Comment by sidorares** [comment]
yeah looks like it is listening ( 33060 is for alternative wire protocol - X Protocol )

can you telnet into 127.0.0.1:3306? `ECONNREFUSED` is usually when node tcp client can't connect to ip/port

**Reply by ghost** [reply]
This is what I get : 

![image](https://user-images.githubusercontent.com/86203108/165441333-924f639b-b1ba-49db-b18f-41c22a3404b3.png)

**Reply by sidorares** [reply]
try `telnet 127.0.0.1 3306`

**Reply by ghost** [reply]
Oh yeah my bad, sorry

![image](https://user-images.githubusercontent.com/86203108/165441894-d138339d-fdef-414f-ae70-a63ec786f57e.png)

**Reply by sidorares** [reply]
do you also use `127.0.0.1` in your node script as `host` parameter in `createConnection` config? ( the error says non-localhost ip address, so the actual problem is that you can't reach that ip + port )

**Reply by ghost** [reply]
No, I use the database server ip, since the node server and the database server are two different servers

**Reply by sidorares** [reply]
can you try to telnet to the database ip+port from the node server host?

**Reply by ghost** [reply]
Node : 

```
import { Telnet } from "telnet-client";

const connection = new Telnet()

const params = {
  host: "dbIP",
  port: 3306,
  shellPrompt: "/ # ",
  timeout: 1500
}

connection.on("ready", prompt => {
	connection.exec(cmd, (error, response) => {
		if (error) {
			console.log(error)
		}

		else {
			console.log(response)
		}
	})
})

connection.on("timeout", () => {
	console.log("socket timeout !")
	connection.end()
})

connection.on("close", () => {
	console.log("connection closed")
})

connection.connect(params)
```

error : 

![image](https://user-images.githubusercontent.com/86203108/165447268-bf745cf5-c06f-4a05-b3d9-c08b575d899d.png)

**Reply by sidorares** [reply]
still the same advise - check node server <-> db server connectivity ( firewall rules etc )

**Reply by ghost** [reply]
Uhm weird, there is no firewall rules for now, and I can connect to the server with websockets on port 15000n but can't for the db on port 3306

**Comment by ghost** [comment, marked as answer]
Yoo I found from were it was coming. By default, mysql server are only open for localhost connection, so you have to change this option.

Here is a tutorial : https://www.digitalocean.com/community/tutorials/how-to-allow-remote-access-to-mysql

Anyways thanks for helping

**Reply by sidorares** [reply]
yeah it has to listen on public interface to be available
